### PR TITLE
Uoe 5310 code review

### PR DESCRIPTION
**Pull request by pm-viral-vala**

### CODE REVIEW

File: endpoints/openrtb2/ctv/impressions/impressions.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+// IImpressions ...
+type IImpressions interface {
+	Get() [][2]int64
+	Algorithm() int // returns algorithm used for computing number of impressions
+}
+
+// NewImpressions generate object of impression generator
+// based on input algorithm type
+func NewImpressions(podMinDuration, podMaxDuration int64, vPod openrtb_ext.VideoAdPod, algorithm int) (IImpressions, error) {
```
Comment (pm-viral-vala) [inline endpoints/openrtb2/ctv/impressions/impressions.go:41]: pass pointer to vPod

File: endpoints/openrtb2/ctv/impressions/impressions.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	switch algorithm {
+	case MaximizeForDuration:
+		ctv.Logf("Selected 'MaximizeForDuration' ")
+		g := newImpGenA1(podMinDuration, podMaxDuration, vPod)
```
Comment (pm-viral-vala) [inline endpoints/openrtb2/ctv/impressions/impressions.go:45]: renaming function with algorithm names

File: endpoints/openrtb2/ctv/impressions/impressions.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+		return &g, nil
+
+	case OptimizeForFillRate:
+		ctv.Logf("Selected 'OptimizeForFillRate' ")
+		g := newImpGenA2(podMinDuration, podMaxDuration, vPod)
```
Comment (pm-viral-vala) [inline endpoints/openrtb2/ctv/impressions/impressions.go:50]: renaming function with an algorithm names

File: endpoints/openrtb2/ctv/impressions/impressions.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	MaximizeForDuration = iota
+	// OptimizeForFillRate algorithm ensures all possible impression breaks are plotted by considering
+	// minimum as well as maxmimum durations and ads received in the ad pod request.
+	// It computes number of impressions with following steps
+	//  1. Passes input configuration as it is (Equivalent of MaximizeForDuration algorithm)
+	//	2. Ad Pod Duration = Ad Pod Max Duration, Number of Ads = max ads
+	//	3. Ad Pod Duration = Ad Pod Max Duration, Number of Ads = min ads
+	//	4. Ad Pod Duration = Ad Pod Min Duration, Number of Ads = max ads
+	//	5. Ad Pod Duration = Ad Pod Min Duration, Number of Ads = min ads
+	OptimizeForFillRate
```
Comment (pm-viral-vala) [inline endpoints/openrtb2/ctv/impressions/impressions.go:25]: I need to check for the relevant names. MinMaxAlgorithm

File: endpoints/openrtb2/ctv/impressions/impressions.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+	"errors"
+
+	"github.com/PubMatic-OpenWrap/prebid-server/endpoints/openrtb2/ctv"
+	"github.com/PubMatic-OpenWrap/prebid-server/openrtb_ext"
+)
+
+const (
+	// MaximizeForDuration algorithm tends towards Ad Pod Maximum Duration, Ad Slot Maximum Duration
+	// and Maximum number of Ads. Accordingly it computes the number of impressions
+	MaximizeForDuration = iota
```
Comment (pm-viral-vala) [inline endpoints/openrtb2/ctv/impressions/impressions.go:16]: Define new type for ImpressionAlgorithm
type ImpressionAlgorithm int
const (
 MaximizeForDuration = iota
 MinMaxAlgo ImpressionAlgorithm = 1
)

File: endpoints/openrtb2/ctv/impressions/impressions.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import (
+	"errors"
+
+	"github.com/PubMatic-OpenWrap/prebid-server/endpoints/openrtb2/ctv"
+	"github.com/PubMatic-OpenWrap/prebid-server/openrtb_ext"
+)
+
+const (
+	// MaximizeForDuration algorithm tends towards Ad Pod Maximum Duration, Ad Slot Maximum Duration
+	// and Maximum number of Ads. Accordingly it computes the number of impressions
```
Comment (pm-viral-vala) [inline endpoints/openrtb2/ctv/impressions/impressions.go:15]: Add Default Algorithm
it returns with AdSlots Min(multiple of 5 not considered)

File: endpoints/openrtb2/ctv/impressions/impressions.go [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+		ctv.Logf("Selected 'MaximizeForDuration' ")
+		g := newImpGenA1(podMinDuration, podMaxDuration, vPod)
+		return &g, nil
+
+	case OptimizeForFillRate:
+		ctv.Logf("Selected 'OptimizeForFillRate' ")
+		g := newImpGenA2(podMinDuration, podMaxDuration, vPod)
+		return &g, nil
+	}
+	return nil, errors.New("Invalid algorithm value")
```
Comment (pm-viral-vala) [inline endpoints/openrtb2/ctv/impressions/impressions.go:53]: return DefaultAlgorithm
and remove returning of error object

File: endpoints/openrtb2/ctv/impressions/impression_generator.go [hunk 8]
```diff
@@ -127,50 +70,50 @@ func (cfg adPodConfig) getImpressions() (adPodConfig, [][2]int64) {
...
+		ctv.Logf("Either cfg.slotMaxDuration or cfg.slotMinDuration or both are <= 0. Hence, totalAds = 0")
 		return 0
 	}
 	maxAds := cfg.podMaxDuration / cfg.slotMaxDuration
 	minAds := cfg.podMaxDuration / cfg.slotMinDuration
 
-	Logf("Computed minAds = %v , maxAds = %v\n", minAds, maxAds)
+	ctv.Logf("Computed minAds = %v , maxAds = %v\n", minAds, maxAds)
 
 	totalAds := max(minAds, maxAds)
```
Comment (pm-viral-vala) [inline endpoints/openrtb2/ctv/impressions/impression_generator.go:105]: check if we can remove max function. as cfg.podMaxDuration / cfg.slotMinDuration condition always returns max value.

File: endpoints/openrtb2/ctv/impressions/helper.go [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+// If computed factor < MinDuration then it will ensure and return
+// close factor >=  MinDuration
+func getClosetFactorForMinDuration(MinDuration int64, multipleOf int64) int64 {
+	closedMinDuration := getClosetFactor(MinDuration, multipleOf)
+
+	if closedMinDuration == 0 {
+		return multipleOf
+	}
+
+	if closedMinDuration == MinDuration {
```
Comment (pm-viral-vala) [inline endpoints/openrtb2/ctv/impressions/helper.go:76]: can we remove this condition?

**Comment by ShriprasadM** [comment]
Changes to be done as per call with @pm-viral-vala 

1. fillZeroSlotsOnPriority - Check if it is required as per new code
2. Algorithm 2 - To make 2 for loops,  in case of go routine
3. maxExpectedDurationMap - to use key as string.  e.g. "4,17"
4. Give name to algo
5. maxAds := cfg.podMaxDuration / cfg.slotMaxDuration
	minAds := cfg.podMaxDuration / cfg.slotMinDuration

	Reverse variable names

6. max(minAds, maxAds) - check if we can remove
7. impressions.NewImpressions - return default algorithm which will have pod request level no of min ads and  ad min duration and ad max duration

File: endpoints/openrtb2/ctv/impressions/impression_generator.go (hunk 8 above)
Comment (ShriprasadM) [reply endpoints/openrtb2/ctv/impressions/impression_generator.go:105]: This is not removed yet.

**Comment by ShriprasadM** [comment]
1. optimizeRepeatations- rename it to getMaximumRepeatations
2.  map[string][2]*int - remove pointer maxExpectedDurationMap
3. rename mainstorage = current counter and temporary storage = max counter
4. Check if we can increment directly current counter instead of keeping max counter
5. MinMaxAlgorithm    -->  not needed Algorithm = iota -  check
6. we can merge func (config *generator) getImpressions() [][2]int64  it to func (config *generator) Get() [][2]int64 {
7. Closet - rename - type here
8. impression_generator_A1.go - rename both files with
actual algorithm name

**Pull request merged**